Мордвинцева, HW3-4

mordvintseva opened the pull request:
Ячейки таблицы пока отображаются не очень корректно при скролле или поиске, постараюсь поправить

Commit:
First homework. Initial commit

Commit:
Add .gitignore

Commit:
Code refactoring

Commit:
Change of characters loading

Character loading when reaching the end of UITableView
Delete initializer of Character struct

Commit:
Delete variable category from CharacterViewController

Commit:
Clear repository on new branch

Commit:
Init commit

Commit:
Database bug fix

Change DBService.delete(_ note: Note) method

Commit:
Revert "Database bug fix"

This reverts commit 826a339695152b8caa0a6e09cfc90fbf57f07a3b.

Commit:
Database bug fix

Change DBService.change(_ note: Note)

Commit:
Fix images display

Add new cell class for notes without image
Add alert when on of fields is empty in AddNoteViewController
Fix search bar

MortyMerr commented:
Когда запускаешь pod install в консоли варнинг об отсутствии версии. Мелочь, но ревьювера бы смутило (детская ошибка)

MortyMerr commented:
Супер сексуально, молодец. Все работает. Иду смотреть код. Я бы еще сделал large titles на экране с табличкой, это тренд. Ну и кнопки Choose image, save, cancel я бы закрепил снизу, а то немного некрасиво смотрится когда они внизу ползут, если текст пишешь
Если это сделаешь - я бы прям выложил куда нибудь в профиль к себе

M0rtyMerr commented on Mordvintseva/Notes/Notes/Extensions/String.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+extension String {
```
Код конечно получился очень красивый, но кидать делать такие extension для общих типов - плохо. Представь сколько там всякого мусора будет, и каждый будет его видеть
Более правильный путь, добавить к каждой ячейке вычислмое поле reuseIdentifier и вычислять его как String(describing: self)
Подробнее здесь - https://github.com/AliSoftware/Reusable. Там всего пять файлов, посмотри код исходников
Советую один раз сделать самой, осознать, потом просто это пользовать

M0rtyMerr commented on Mordvintseva/Notes/Notes/Extensions/UIImageView.swift:
```diff
@@ -0,0 +1,33 @@
+
+import UIKit
+
+extension UIImageView {
```
И снова красиво (работает, удобно), но не по правилам промышленного программирование. Представь, что завтра придет начальник и скажет - лол, а я хочу чтобы ты их сохраняла в базу данных. Или отсылала на удаленный сервер по интернету. Придется прям переписывать.
Правильно было бы сделать протокол `ImageRepository` и сокрыть за ним протоколом детали реализации. Ну и у него были бы методы save(image: UIImage)
Тогда ты в любой момент могла бы внутри поменять куда сохранять, не меняя остальной код

M0rtyMerr commented on Mordvintseva/Notes/Notes/Extensions/UITableView.swift:
```diff
@@ -0,0 +1,31 @@
+//
+//  UITableView.swift
```
Не называй файлы с экстеншнами в честь класса. Либо
UITableView+setEmptyView.swift (это если одна функция на файл)
Либо UITableViewUtils
Если называешь классом - путает

M0rtyMerr commented on Mordvintseva/Notes/Notes/Extensions/UIImageView.swift:
```diff
@@ -0,0 +1,33 @@
+        return imagePath
+    }
+
+    func getImage(path: String) {
```
Метод getImage, который ничего не возвращает - плохо. В данном варианте это скорее loadImage

M0rtyMerr commented on Mordvintseva/Notes/Notes/Services/DBService.swift:
```diff
@@ -0,0 +1,73 @@
+        return transformResultsToArray(result: result)
+    }
+
+    private func transformResultsToArray(result: Results<Note>) -> [Note] {
```
Блэээээээээт, сколько еще раз я должен пошутить, чтобы метод array.append(contentsOf:) запомнили все

M0rtyMerr replied:
Можно просто  return Array(result)

M0rtyMerr replied:
Пойду разбираться но пока не очень понял, зачем тебе массив. Result в Realm самообновляемый. То есть если заметка добавляется, она подтянется в Result. Зачем убирать это поведение и кастить в массив - не понял

mordvintseva replied:
В целом, я помню про append, но когда я попыталась сделать append(consistOf: ...) на меня ругался компилятор из-за того, что работаю с Results.
В массив переписываю, чтобы скрыть от ViewController'a работу с типами базы данных, потому что если бы я возвращала Results<Note> мне пришлось бы делать import Realm.
Можно сделать как-то иначе?

M0rtyMerr replied:
Ну... Это спорно. В принципе согласен. С другой стороны раз уж есть автообновляемый тип, почему бы им не пользоваться

M0rtyMerr replied:
Оставь

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/ViewModel/NotesListViewModelProtocol.swift:
```diff
@@ -0,0 +1,22 @@
+protocol NotesListViewModelProtocol {
+    var onNotesChanged: (([Note]) -> Void)? { get set }
+
+    init (database: DBService)
```
Мне не нравится. Протокол как раз для того и нужен чтобы сокрыть детали реализации. Зачем пихать в него конструктор? А вдруг завтра из нетворка?

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/ViewModel/NotesListViewModelProtocol.swift:
```diff
@@ -0,0 +1,22 @@
+    init (database: DBService)
+
+    func load()
+    func add(_ data: [String: String])
```
А вот это [String: String] точно никак нельзя в Note замапить?

M0rtyMerr commented on Mordvintseva/Notes/Notes/Extensions/Notification.Name.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+extension Notification.Name {
```
Уважаю

M0rtyMerr commented on Mordvintseva/Notes/Notes/Extensions/UITableView.swift:
```diff
@@ -0,0 +1,31 @@
+
+extension UITableView {
+    func setEmptyView(title: String) {
+        let emptyView = UIView(frame: CGRect(x: self.center.x, y: self.center.y, width: self.bounds.size.width, height: self.bounds.size.height))
```
Норм, но тебе самой то удобно? А если завтра будет два варианта EmptyView? Или три разных emptyView для трех разных таблиц? будет setEmptyViewOne, setEmptyViewTwo, setEmptyViewThree
К тому же каждый раз создается новая вьюха
Как вариант можно сверстать emptyView в отдельном xib. И сдкелать отдельный класс. Что думаешь?

mordvintseva replied:
ctrl +c, ctrl + v было не пыльно, но я поняла)

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/NotesListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+
+    @objc func onDidReceiveData(_ notification: Notification) {
+        print("Data did receive")
+        if let data = notification.userInfo as? [String: String] {
```
NotificationCenter плохой способ общаться. Об изменениях знает все приложение. Молодец, что попробовала. Но может есть другие идеи? Через замыкания?

M0rtyMerr replied:
Посмотри coordinator pattern (скинул вк). Переписывать не заставляю, но вдруг заинтересует

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/NotesListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+
+extension NotesListViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        if notes.isEmpty == true {
```
А точно ли нужно сравнивать с true?

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/NotesListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+                fatalError("Cell can not be displayed")
+            }
+
+            cell.setInfo(note: notes[indexPath.row])
```
А если закрыть ячейки сверху протоколом где есть один метод `setInfo(note:)`, можно сократить это if else в два раза

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/NotesListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
+        if editingStyle == .delete {
```
guard бы подошел лучше, как думаешь?

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/NotesListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
+        if editingStyle == .delete {
+            viewModel.delete(notes[indexPath.row])
+            tableView.beginUpdates()
```
Зачем используем beginUpdates?

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/ViewModel/NotesListViewModel.swift:
```diff
@@ -0,0 +1,57 @@
+
+    func add(_ data: [String: String]) {
+        let note = Note(data: data)
+        database.add(note)
```
Если ооооочень хочется оставить [String: String] (хотя никто не будет понимать какие там ключи), то нужно здесь переиспользовать код и написать `self.add(note)

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -0,0 +1,95 @@
+        var finalNotes = [Note]()
+        let record = Note(data: note)
+
+        viewModel.deleteAll()
```
Вот эту очистку базы данных я бы убрал в setup

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -0,0 +1,95 @@
+        }
+        viewModel.search("nope")
+
+        XCTAssertEqual(initNotes.count, 1)
```
Точно нельзя этот тест покороче переписать? (коммент не к конкретной строке а ко всему вообще)

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -0,0 +1,95 @@
+        }
+        viewModel.delete(record)
+
+        XCTAssertTrue(finalValue == initValue - 1)
```
AssertEqual

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -0,0 +1,95 @@
+        viewModel.load()
+
+        XCTAssertNotEqual(initValue, finalValue)
+        viewModel.delete(Note(data: note))
```
А это маленькое удаление тут что забыло?)

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -0,0 +1,95 @@
+
+    override func setUp() {
+        super.setUp()
+        viewModel = NotesListViewModel(database: DBService())
```
Вот прям если по хорошему, то надо заменять DBService на мок. А в моке будет обычный массив. Сама заметила как неудобно, что у тебя состояние хранится и его надо чистить

MortyMerr commented:
Основное нарекание к тесту на реальной базе и к extension UIImageView вместо сервиса. Пойду ка я заберу плашку сделанного задания, надо бы еще поработать

Commit:
Update pods

Commit:
Cell modification

Add pod Reusable
Add CellProtocol

Commit:
Add ImageService

Add ImageService instead UIImageView extension

Commit:
Change EmptyView using

Delete setEmptyView from UITableView extension
Create new Xib and class files with EmptyView

Commit:
Rename UITableView extension file

Commit:
Fix keyboard view

Fix overlapping of elements by keyboard
Add "Done" button to keyboard when typing

Commit:
Code correction

Commit:
Delete NotificationCenter usage

Remove data transfer through NotificationCenter
Add onAddNote completionHandler to AddNoteViewController

Commit:
Fix protocols

Commit:
Update tests

Commit:
Make viewControllers and views reusable

mordvintseva commented:
Исправила замечания, надеюсь, тесты я правильно поняла
По поводу кнопок, которые ползут, когда набираешь. Тогда получается, скролл теряет смысл, если их закреплять снизу? TextView будет определенной высоты, а сейчас он меняет высоту в зависимости от набранного текста.

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/AddNote/AddNoteViewController.swift:
```diff
@@ -39,23 +42,62 @@ class AddNoteViewController: UIViewController, UINavigationControllerDelegate {
         var data: [String: String] = [:]
-        if let imagePath = imageView.saveImage() {
+        if let image = imageView.image {
+            let imagePath = ImageRepository().save(image: image)
```
А может ImageRepository сделать создать один раз и сделать полем?

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/AddNote/AddNoteViewController.swift:
```diff
@@ -39,23 +42,62 @@ class AddNoteViewController: UIViewController, UINavigationControllerDelegate {
         noteView.delegate = self
+        createToolbar()
+
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboard), name: UIResponder.keyboardWillHideNotification, object: nil)
```
А зачем сразу две?

mordvintseva replied:
keyboardWillChangeFrameNotification увеличивает экран снизу на высоту клавиатуры, чтобы было возможно пролистать до нижних кнопок, не закрывая клавы.
keyboardWillHideNotification возвращает в прежнее состояние, иначе экран растягивается, появляется ненужное пустое пространство снизу под кнопками

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/AddNote/AddNoteViewController.swift:
```diff
@@ -39,23 +42,62 @@ class AddNoteViewController: UIViewController, UINavigationControllerDelegate {
+
+    func createToolbar() {
+        let toolBar = UIToolbar()
+        toolBar.sizeToFit()
```
Уверена, что строчка нужная?

mordvintseva replied:
Без нее Done не отображается

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/AddNote/AddNoteViewController.swift:
```diff
@@ -39,23 +42,62 @@ class AddNoteViewController: UIViewController, UINavigationControllerDelegate {
+            scrollView.contentInset = UIEdgeInsets(top: 0, left: 0, bottom: keyboardViewEndFrame.height, right: 0)
+        }
+
+        scrollView.scrollIndicatorInsets = scrollView.contentInset
```
А я забываю про это всегда, красава :)

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/Cells/CellProtocol.swift:
```diff
@@ -0,0 +1,15 @@
+import Reusable
+import UIKit
+
+protocol CellProtocol: NibReusable {
```
А зачем?

mordvintseva replied:
Мне показалось, так должно выглядеть исправление на твой комментарий:

> А если закрыть ячейки сверху протоколом где есть один метод setInfo(note:), можно сократить это if else в два раза

M0rtyMerr commented on Mordvintseva/Notes/Notes/Controllers/NotesList/NotesListViewController.swift:
```diff
@@ -52,56 +52,43 @@ class NotesListViewController: UIViewController, UITableViewDelegate {
+
+        cell.setInfo(note: notes[indexPath.row])
+        // swiftlint:disable:next force_cast
+        return cell as! UITableViewCell
```
А зачем?

M0rtyMerr commented on Mordvintseva/Notes/Notes/Services/DBServiceRealm.swift:
```diff
@@ -19,7 +19,7 @@ class DBService {
     func getAll() -> [Note] {
         let result = realm.objects(Note.self)
-        return transformResultsToArray(result: result)
+        return [Note](result)
```
Магия

M0rtyMerr replied:
А точно по другому нельзя?

mordvintseva replied:
> Магия

Сарказм?(

M0rtyMerr replied:
Неа. Я реально не видел такого способа создания массива. ИМХО, `Array(result)` лаконичнее

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -12,11 +12,17 @@ import XCTest
         super.setUp()
-        viewModel = NotesListViewModel(database: DBService())
+        viewModel = NotesListViewModel(database: MockDBService())
+        notes.forEach { viewModel.add($0) }
```
Можно `notes.forEach(viewModel.add)`

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -25,71 +31,108 @@ class NotesTests: XCTestCase {
     }
 
     func testOnNotesChanged() {
+        // check onNotesChangedCalled
```
Лучше такие комменты пихать в название теста

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -25,71 +31,108 @@ class NotesTests: XCTestCase {
     }
 
     func testAddNote() {
+        // add new note and compare size of notes data
```
Комменты зло, код должен выражать свою суть

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -25,71 +31,108 @@ class NotesTests: XCTestCase {
+        XCTAssertEqual(initValue, finalValue - 1)
+
+        // delete test note from DB
+        viewModel.delete(note)
```
Почему не deleteAll() в tearDown()?

mordvintseva replied:
Не хотела, чтобы эта тестовая запись фигурировала в остальных тестах. Можно задать супер уникальные данные, конечно, и не удалять

M0rtyMerr replied:
Так почему базу не чистить после каждого теста?

mordvintseva replied:
А, всё, поняла, думала, что tearDown и setUp только один раз вызываются за всё тестирование

M0rtyMerr replied:
В tearDown еще надо очистить внутренние поля

mordvintseva replied:
Внутренние поля записей в базе?

M0rtyMerr replied:
Внутренние поля класса private var viewModel

mordvintseva replied:
Исправила, но поскольку база данных в модели - приватное поле, пришлось создавать наследуемый класс и там объявлять ее public полем.
В целом, я посмотрела, база деинититься и через просто viewModel = nil, может, создание наследуемого класса было излишним?

MortyMerr commented:
Молодец, все ок

Commit:
Code refactoring

Commit:
Change keyboard observing

Change keyboardWillChangeFrameNotification to keyboardWillShowNotification
Pin buttons to bottom in AddNoteViewController

Commit:
Test fix

M0rtyMerr commented on Mordvintseva/Notes/NotesTests/NotesTests.swift:
```diff
@@ -21,17 +21,18 @@ class NotesTests: XCTestCase {
     override func setUp() {
         super.setUp()
-        viewModel = NotesListViewModel(database: MockDBService())
+        viewModel = NotesListViewModelTest(database: MockDBService())
```
Просто сохранит MockDBService() в поле тестового класса и вызывай clear в tearDown и будет пушка
